---
title: "HISAT OMIXON comparison"
output: html_document
---

- OMIXON also called DPA, DPB 
- Some OMIXON alleles were called as "none [identical alleles" - genotypes with 1 or 2 missing alleles were omitted from the comparison
- Alelle expression suffixes such as "LQ", "Q", "N" etc were stripped of before comparisons were made
- Alleles did not always match in terms of number of subtypes called. In this case, they were matched up until the last common subtype, e.g. "HISAT:C07:02:01 == OMIXON:C07:02". Note that 
 however e.g. "HISAT:C07:02:01 != OMIXON:C07:02:03" , i.e. this is classed as a mismatch. 
- The "matches" output sheet lists samples in rows, and has 6 columns, 1 for each HLA genotype. Column values are 0, 1, 2 or NA, where 0 indicates no allele matches, 1 indicates 1 allele match, 2 indicates 2 allele matches, and NA means that the comparison was not done due to one or more missing OMIXON alleles.
- The "mismatches" output sheet is a log of all mismatch occurences

## Summary of percentage matches

The columns 0, 1 or 2 summarizes the number of samples with 0 alleles that matched in their HISAT and OMIXON calls, the number of samples with 1 matching allele, and the number of samples with 2 matching alleles. E.g. for HLA-A, there were 19 samples that had no matching alleles, 223 samples with 1 matching allele, and 629 samples that had 2 matching alleles. % matches denotes the percentage of alleles that matched (e.g. (223 + 629\*2) / ( (19 + 223+ 629)\*2) ) = 0.8502)

```{r, echo=FALSE}

library(stringr)

omixon.raw <- read.delim("data/raw/OMIXON_HLAoutput.txt", head=F, stringsAsFactors = F)
hisat.raw <- read.delim("data/raw/HISAT_CAAPA_HLA_2.txt", head=T, stringsAsFactors = F)

getHiSatAlleles <- function(sample_id, genotype) {
  alleles <- hisat.raw[(hisat.raw$Genome.ID == sample_id),2]
  alleles <- alleles[grep(paste0("^", genotype), alleles)]
  alleles <- gsub("[A-Z]$", "", alleles)
  alleles <- gsub(paste0(genotype, "\\*"), "", alleles)
  return (alleles[order(alleles)])
}

getOmixonAlleles <- function(sample_id, genotype) {
  alleles <- str_trim(unlist(omixon.raw[(omixon.raw$V1 == sample_id),-c(1:2)]))
  alleles <- alleles[grep(paste0("^", genotype), alleles)]
  if (length(alleles) > 0) {
    alleles <- str_trim(unlist(str_split(alleles, "\\["))[seq(1,length(alleles)*2,2)])
    alleles <- gsub("\\*LQ", "", alleles)
    alleles <- gsub("[A-Z]$", "", alleles)
    alleles <- gsub(genotype, "", alleles)
    return (alleles[order(alleles)])
  } else {
    return (alleles)
  }
}

compareAlleles <- function(sample_id, genotype) {
  hisat.alleles <- getHiSatAlleles(sample_id, genotype) 
  omixon.alleles <- getOmixonAlleles(sample_id, genotype)
  
  #Only compare genotypes where both OMIXON alleles were called
  if (length(omixon.alleles) != 2) { 
    return (NA)
  }
  
  #Count the nr of matches
  nr.matches <- 0
  if (grepl(omixon.alleles[1], hisat.alleles[1]) | grepl(hisat.alleles[1], omixon.alleles[1])) {
    nr.matches <- nr.matches + 1
  }
  if (grepl(omixon.alleles[2], hisat.alleles[2]) | grepl(hisat.alleles[2], omixon.alleles[2])) {
    nr.matches <- nr.matches + 1
  }
  #If there are no matches, 1 allele might still match, of one changes the ordering
  #This caters for the case with HISAT = 01:01:01:01;33:03:01, OMIXON = 33:03:01;36:01
  if (nr.matches == 0) {
    if (grepl(omixon.alleles[1], hisat.alleles[2]) | grepl(hisat.alleles[1], omixon.alleles[2]) |
        grepl(omixon.alleles[2], hisat.alleles[1]) | grepl(hisat.alleles[2], omixon.alleles[1])) {
      nr.matches <- nr.matches + 1
    }
  }
  
  #Log the mismatches
  if (nr.matches != 2) {
    cat(sample_id, paste(genotype, hisat.alleles, sep=":"), 
          paste(genotype, omixon.alleles, sep=":"),
          nr.matches, "\n",
          sep="\t", append=T, file="data/output/mismatches.txt")
  }
  
  return (nr.matches)
}

out.frame <- data.frame()
cat("sample_id", "hisat_a1", "hisat_a2", "omixon_a1", "omixon_a2", "nr_matches", "\n",
    sep="\t", append=F, file="data/output/mismatches.txt")
for (sample_id in unique(hisat.raw$Genome.ID)) {
  out.frame <- rbind(out.frame, 
                     data.frame(
                       sample_id=sample_id,
                       "A"=compareAlleles(sample_id, "A"),
                       "B"=compareAlleles(sample_id, "B"),
                       "C"=compareAlleles(sample_id, "C"),
                       "DQA1"=compareAlleles(sample_id, "DQA1"),
                       "DQB1"=compareAlleles(sample_id, "DQB1"),
                       "DRB1"=compareAlleles(sample_id, "DRB1")))
}
write.table(out.frame, "data/output/matches.txt",
            sep="\t", quote=F, row.names=F, col.names=T)

summary.frame <- data.frame(rbind(table(out.frame$A),
                               table(out.frame$B),
                               table(out.frame$C),
                               table(out.frame$DQA1),
                               table(out.frame$DQB1)))
last.row <- c(table(out.frame$DRB1),0)
names(last.row)[3] <- c("X2")
summary.frame <- rbind(summary.frame, last.row)
colnames(summary.frame) <- c("0", "1", "2")
rownames(summary.frame) <- c("HLA-A", "HLA-B", "HLA-C", "DQA1", "DQB1", "DRB1")

summary.frame$perc.matches <- NA
for (i in 1:dim(summary.frame)[1]) {
  summary.frame$perc.matches[i] <- 
    round((summary.frame[i,2] + 2*summary.frame[i,3])/(sum(summary.frame[i,], na.rm=T)*2)*100,2)
}

colnames(summary.frame)[4] <- "% matches"

knitr::kable(summary.frame)

```

