# hisat2 
HISAT2 is a fast and sensitive alignment program for mapping next-generation sequencing reads (whole-genome, transcriptome, and exome sequencing data) against general population (as well as a single referece genome). Based on generalized BWT (Sir√©n et al 2014 - Indexing graphs for path queries with applications in genome research), we designed and implemented generalized FM index (GFM), which is the first implementation of GFM to the best of our knowledge, and in addition to one global GFM index that represents general population, HISAT2 uses a large set of small GFM indexes that collectively cover the whole genome (each index represents a genomic region of ~57,000 bp and ~54,000 indexes are needed to cover human population). These small indexes (called local indexes) combined with several alignment strategies enable effective alignment of sequencing reads. This new indexing scheme is called Hierarchical Generalized FM index (HGFM).  The memory footprint of HISAT2 is relatively low (~7GB for human population). We have developed HISAT2 based on the HISAT and Bowtie2 implementation.

We plan to release an alpha version of HISAT2 in August this year.
